modular_zubbers/code/modules/rimpoint_newfeatures/code/umbrella.dm

```diff
@@ -0,0 +1,146 @@
+	if(!looping_sound)
+		looping_sound = new(src, FALSE)
+	if(!user_holding)
+		RegisterSignal(user, COMSIG_MOVABLE_MOVED, PROC_REF(on_move))
+	user_holding = user
+	proccess_rainsounds()
+
+/obj/item/umbrella/dropped(mob/user, silent)
+	. = ..()
+	revoke_trait(user, immunity_type)
+	QDEL_NULL(looping_sound)
+	UnregisterSignal(user, COMSIG_MOVABLE_MOVED)
+	user_holding = null
+
+/obj/item/umbrella/proc/on_move(atom/old_loc, dir, forced, list/old_locs)
+	SIGNAL_HANDLER
+
+	proccess_rainsounds(old_loc)
+
+/obj/item/umbrella/proc/proccess_rainsounds()
+	if(!SSweather)
+		return
+
+
+	var/turf/turf = get_turf(user_holding)
+	if(!turf)
+		return FALSE
+	var/datum/weather/weather = SSweather.get_weather(turf.z, get_area(turf))
+	if(!weather || !(turf.loc in weather.impacted_areas) )
+		looping_sound.stop() // If there is no weather, why even have a sound?
+		return FALSE
+	if(weather.recursive_weather_protection_check(user_holding)) // If the player is protected by the weather, play the umbrella sounds.
+		looping_sound.start()
+	else
+		looping_sound.stop()
+
+/obj/item/umbrella/collapsible
+	name = "collapsible umbrella"
+	icon = 'modular_zubbers/icons/obj/equipment/umbrella.dmi'
+	icon_state = "umbrella_retract_off"
+	slot_flags = ITEM_SLOT_POCKETS
+	w_class = WEIGHT_CLASS_SMALL
+	inhand_icon_state = "sheathed_umbrella"
+
+	/// Whether the umbrella is currently collapsed or open.
+	var/collapsed = FALSE
+	/// sound when toggled open/closed
+	var/toggle_sound = 'modular_zubbers/code/modules/rimpoint_newfeatures/sound/effects/umbrella_open.ogg'
+	var/close_sound = 'modular_zubbers/code/modules/rimpoint_newfeatures/sound/effects/umbrella_close.ogg'
+
+/obj/item/umbrella/collapsible/Initialize(mapload)
+	. = ..()
+	AddComponent( \
+		/datum/component/transforming, \
+		w_class_on = WEIGHT_CLASS_BULKY, \
+		attack_verb_continuous_on = list("whaps", "thwacks", "whacks", "beats"), \
+		attack_verb_simple_on = list("whap", "thwack", "whack", "beat"), \
+	)
+	RegisterSignal(src, COMSIG_TRANSFORMING_ON_TRANSFORM, PROC_REF(on_transform))
+
+
+/obj/item/umbrella/collapsible/proc/on_transform(obj/item/source, mob/user, active)
+	SIGNAL_HANDLER
+
+	collapsed = active
+	if(!user)
+		return FALSE
+	if(collapsed)
+		icon_state = "umbrella_retract_on"
+		give_trait(user, immunity_type)
+		inhand_icon_state = "umbrella"
+		slot_flags = NONE
+		if(toggle_sound)
+			playsound(user, toggle_sound, 100)
+		to_chat(user, "You open the umbrella.")
+	else
+		icon_state = "umbrella_retract_off"
+		inhand_icon_state = "sheathed_umbrella"
+		slot_flags = ITEM_SLOT_POCKETS
+		revoke_trait(user, immunity_type)
+		if(toggle_sound)
+			playsound(user, close_sound, 100)
+		to_chat(user, "You close the umbrella.")
+
+	balloon_alert(user, active ? "extended" : "collapsed")
+	proccess_rainsounds()
+	return COMPONENT_NO_DEFAULT_MESSAGE
+
+
+
+/obj/item/umbrella/proc/give_trait(mob/living/user,immunity_type)
+	if(!HAS_TRAIT(user, immunity_type))
+		ADD_TRAIT(user, immunity_type, REF(src))
```

Comment:
missing space between procs, also duplicate traits of the same source are not possible, has traits are not neccesary